File: services/engine/src/evidra/api/conversations.py
from fastapi import APIRouter, Query, Request
from fastapi.responses import Response

from evidra.api.documents import services
from evidra.conversations.models import (
    ConversationCreate,
    ConversationPage,
    ConversationRecord,
    RunPage,
    RunPrepare,
    RunRecord,
    VectorBuild,
    VectorJob,
)

router = APIRouter(
    prefix="/v1/notebooks/{notebook_id}/snapshots/{snapshot_id}", tags=["conversations"]
)


@router.post("/conversations", response_model=ConversationRecord, status_code=201)
def create(
    notebook_id: str, snapshot_id: str, body: ConversationCreate, request: Request
) -> ConversationRecord:
    service, context = services(request, notebook_id, snapshot_id, commit=True)
    return service.conversations.create(context, body)


@router.get("/conversations", response_model=ConversationPage)
def listing(
    notebook_id: str,
    snapshot_id: str,
    request: Request,
    offset: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=50),
) -> ConversationPage:
    service, context = services(request, notebook_id, snapshot_id)
    return service.conversations.list(context, offset, limit)


@router.get("/conversations/{conversation_id}", response_model=ConversationRecord)
def conversation(
    notebook_id: str, snapshot_id: str, conversation_id: str, request: Request
) -> ConversationRecord:
    service, context = services(request, notebook_id, snapshot_id)
    return service.conversations.conversation(context, conversation_id)


@router.get("/conversations/{conversation_id}/runs", response_model=RunPage)
def history(
    notebook_id: str,
    snapshot_id: str,
    conversation_id: str,
    request: Request,
    offset: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=50),
) -> RunPage:
    service, context = services(request, notebook_id, snapshot_id)
    return service.conversations.runs(context, conversation_id, offset, limit)


@router.post("/conversations/{conversation_id}/runs", response_model=RunRecord, status_code=201)
async def prepare(
    notebook_id: str, snapshot_id: str, conversation_id: str, body: RunPrepare, request: Request
) -> RunRecord:
    service, context = services(request, notebook_id, snapshot_id, commit=True)
    return await service.conversations.prepare(context, conversation_id, body)


@router.get("/runs/{run_id}", response_model=RunRecord)
def run(notebook_id: str, snapshot_id: str, run_id: str, request: Request) -> RunRecord:
    service, context = services(request, notebook_id, snapshot_id)
    return service.conversations.read(context, run_id)


@router.post("/runs/{run_id}/start", response_model=RunRecord)
async def start(notebook_id: str, snapshot_id: str, run_id: str, request: Request) -> RunRecord:
    service, context = services(request, notebook_id, snapshot_id, commit=True)
    return service.conversations.start(context, run_id)


@router.post("/runs/{run_id}/cancel", response_model=RunRecord)
async def cancel(notebook_id: str, snapshot_id: str, run_id: str, request: Request) -> RunRecord:
    service, context = services(request, notebook_id, snapshot_id, commit=True)
    return service.conversations.cancel(context, run_id)


@router.get("/runs/{run_id}/events", response_class=Response)
def events(
    notebook_id: str, snapshot_id: str, run_id: str, request: Request, cursor: int = Query(0, ge=0)
) -> Response:
    service, context = services(request, notebook_id, snapshot_id)
    page = service.conversations.events(context, run_id, cursor)
    # Finite SSE batches make native transport bounded and reconnects explicit/no replayed send.
    return Response(
        f"id: {page.cursor}\nevent: batch\ndata: {page.model_dump_json()}\n\n",
        media_type="text/event-stream",
        headers={"Cache-Control": "no-store"},
    )


@router.post("/vectors", response_model=VectorJob, status_code=202)
async def build(
    notebook_id: str, snapshot_id: str, body: VectorBuild, request: Request
) -> VectorJob:
    service, context = services(request, notebook_id, snapshot_id, commit=True)
    return service.vectors.start(context, body)


@router.get("/vectors/{job_id}", response_model=VectorJob)
def vector_job(notebook_id: str, snapshot_id: str, job_id: str, request: Request) -> VectorJob:
    service, context = services(request, notebook_id, snapshot_id)
    return service.vectors.read(context, job_id)


@router.post("/vectors/{job_id}/cancel", response_model=VectorJob)
async def cancel_vector(
    notebook_id: str, snapshot_id: str, job_id: str, request: Request
) -> VectorJob:
    service, context = services(request, notebook_id, snapshot_id, commit=True)
    return service.vectors.read(context, job_id, cancel=True)

File: apps/zotero/src/bridge/conversations.ts
import type { ConversationCommand, ProviderCommand } from './types';
import type { EngineController } from '../bootstrap/engine';

export async function conversationCommand(message: ConversationCommand,
    request: (method: 'GET' | 'POST', path: string, body?: unknown) => Promise<unknown>): Promise<unknown> {
    switch (message.op) {
        case 'conversation.list': return request('GET', `/conversations?offset=${message.offset}&limit=50`);
        case 'conversation.create': return request('POST', '/conversations', message.request);
        case 'conversation.read': return request('GET', `/conversations/${message.conversation_id}`);
        case 'conversation.history': return request('GET', `/conversations/${message.conversation_id}/runs?offset=${message.offset}&limit=50`);
        case 'conversation.prepare': return request('POST', `/conversations/${message.conversation_id}/runs`, message.request);
        case 'conversation.run': return request('GET', `/runs/${message.run_id}`);
        case 'conversation.start': return request('POST', `/runs/${message.run_id}/start`);
        case 'conversation.cancel': return request('POST', `/runs/${message.run_id}/cancel`);
        case 'conversation.events': return request('GET', `/runs/${message.run_id}/events?cursor=${message.cursor}`);
        case 'conversation.vectors.build': return request('POST', '/vectors', message.request);
        case 'conversation.vectors.read': return request('GET', `/vectors/${message.job_id}`);
        case 'conversation.vectors.cancel': return request('POST', `/vectors/${message.job_id}/cancel`);
    }
}

export async function providerCommand(message: ProviderCommand, engine: EngineController): Promise<unknown> {
    switch (message.op) {
        case 'provider.list': return engine.request('GET', `/v1/providers/profiles?offset=${message.offset}&limit=50`);
        case 'provider.write': return engine.request('PUT', `/v1/providers/profiles/${message.profile_id}`, message.request);
        case 'provider.resume': return engine.request('POST', `/v1/providers/profiles/${message.profile_id}/resume`, message.request);
        case 'provider.models': return engine.request('GET', `/v1/providers/profiles/${message.profile_id}/models?offset=${message.offset}&limit=50`);
        case 'provider.settings': return engine.request('GET', '/v1/providers/settings');
        case 'provider.settings.write': return engine.request('PUT', '/v1/providers/settings', message.request);
        case 'provider.secret': return engine.request('GET', `/v1/providers/profiles/${message.profile_id}/secret`);
        case 'provider.secret.write': return engine.request('PUT', `/v1/providers/profiles/${message.profile_id}/secret`, message.request);
        case 'provider.secret.delete': return engine.request('DELETE', `/v1/providers/profiles/${message.profile_id}/secret`, message.request);
        case 'provider.consent': return engine.request('GET', `/v1/notebooks/${message.notebook_id}/providers/${message.profile_id}/consent`);
        case 'provider.consent.write': return engine.request('PUT', `/v1/notebooks/${message.notebook_id}/providers/${message.profile_id}/consent`, message.request);
        case 'provider.price': return engine.request('POST', '/v1/providers/prices', message.request);
        case 'provider.budget': return engine.request('PUT', `/v1/notebooks/${message.notebook_id}/snapshots/${message.snapshot_id}/provider-budgets`, message.request);
        case 'provider.budget.read': return engine.request('GET', `/v1/notebooks/${message.notebook_id}/snapshots/${message.snapshot_id}/provider-budgets/${message.kind}/${message.identity}`);
        case 'provider.calls': return engine.request('GET', `/v1/notebooks/${message.notebook_id}/snapshots/${message.snapshot_id}/provider-calls?offset=${message.offset}&limit=50`);
    }
}
